call-home/src/bin/stats/cache/events_cache.rs

```
@@ -0,0 +1,94 @@
+                key: EVENT_STATS_DATA.to_string(),
+            }
+            .build(),
+        )?;
+
+        let event_set = serde_json::from_str(value)
+            .context(errors::EventSerdeDeserialization { event: value })?;
+        Ok(event_set)
+    }
+
+    fn inc_counter(&mut self, category: Category, action: Action) -> errors::Result<()> {
+        match category {
+            Category::Pool => self.pool.update_counter(action),
+            Category::Volume => self.volume.update_counter(action),
+            Category::Unknown => Ok(()),
+        }
+    }
+}
+
+impl From<&mut EventSet> for EventSet {
+    fn from(event_set: &mut EventSet) -> Self {
+        EventSet {
+            pool: event_set.pool.clone(),
+            volume: event_set.volume.clone(),
+        }
+    }
+}
+
+/// Cache to store data that has to be exposed though exporter.
+pub struct Cache {
+    events: EventSet,
+}
+
+impl Cache {
+    /// Initialize the cache with default value.
+    pub(crate) fn initialize(events: EventSet) {
+        CACHE.get_or_init(|| Mutex::new(Self { events }));
+    }
+
+    /// Returns cache.
+    pub fn get_cache() -> &'static Mutex<Cache> {
+        CACHE.get().expect("Cache is not initialized")
+    }
+
+    /// Get data field in cache.
+    pub fn data_mut(&mut self) -> &mut EventSet {
+        &mut self.events
+    }
+}
+
+/// To store data in shared variable i.e cache.
+pub(crate) async fn store_events(mut sub: BusSubscription<EventMessage>) -> errors::Result<()> {
+    while let Some(message) = sub.next().await {
+        let mut cache = Cache::get_cache().lock().expect("not poisoned");
+        let events_cache = cache.deref_mut();
+        events_cache
+            .data_mut()
+            .inc_counter(message.category, message.action)?;
+    }
+    Ok(())
+}
+
+/// Trait for updating the counters.
+pub(crate) trait StatsCounter {
+    fn update_counter(&mut self, action: Action) -> errors::Result<()>;
```

Does it need to return an error?

---

removed